manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,470 @@
+// Allocator. It exists so that Allocator changes can be abandoned
+// without complex rollback logic. To commit changes to the Allocator, call
+// the proposal's Commit function
+type Proposal interface {
+	// Ports returns the port assignments generated by this proposal
+	Ports() []*api.PortConfig
+
+	// Commit commits the result of the Allocator transaction and alters
+	// the state of the port allocator. Part of the contract of the
+	// Allocator is that it should always return a valid proposal, and so
+	// Commit will always succeed and does not need to return an error.
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify the net
+	// allocator state. IsNoop does not mean the ports for an endpoint haven't
+	// changed. It only means that the publish ports marked in use by the
+	// allocator haven't changed.
+	IsNoop() bool
+}
+
+type proposal struct {
+	pa         *allocator
+	ports      []*api.PortConfig
+	allocate   map[port]struct{}
+	deallocate map[port]struct{}
+}
+
+func (prop *proposal) Ports() []*api.PortConfig {
+	if prop.ports == nil {
+		return []*api.PortConfig{}
+	}
+	return prop.ports
+}
+
+// Commit commits the proposal to the port allocator.
+func (prop *proposal) Commit() {
+	if prop.IsNoop() {
+		// nothing to do if proposal is noop, short circuit a bit
+		return
+	}
+	// The pattern here is we're going to free every port in p.deallocate and
+	// then allocate every port in allocate. any over lap results in no net
+	// change
+	for p := range prop.deallocate {
+		delete(prop.pa.ports, p)
+	}
+	for p := range prop.allocate {
+		prop.pa.ports[p] = struct{}{}
+	}
+}
+
+// IsNoop returns true if the ports in use before this proposal are the same as
+// the ports in use after.
+func (prop *proposal) IsNoop() bool {
+	// if the allocate and deallocate sets are the same length, and every port
+	// in allocate is also found in deallocate, then this proposal is a noop
+	if len(prop.allocate) != len(prop.deallocate) {
+		return false
+	}
+	for p := range prop.allocate {
+		if _, ok := prop.deallocate[p]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// NewAllocator returns a new instance of the Allocator object
+func NewAllocator() Allocator {
+	return &allocator{
+		ports: make(map[port]struct{}),
+	}
+}
+
+// Restore adds the current endpoints to the local state of the port allocator
+// but does not perform any new allocation.
+func (pa *allocator) Restore(endpoints []*api.Endpoint) {
+	// NOTE(dperny) we can be sure that we're not allocating new or conflicting
+	// state because if an endpoint is unallocated, it will not have any ports.
+	// we can't look at the Spec in this method, because the spec isn't real
+	// state, it's just what the user wants. no matter what changes, that
+	// invariant needs to hold.
+
+	// create and commit a proposal for port allocation
+	prop := &proposal{
+		pa: pa,
+		// we only allocate
+		allocate: make(map[port]struct{}),
+	}
+	for _, endpoint := range endpoints {
```

Comment:
We seem to be doing redundant work here but iterating twice, once over the endpoints and then over each of the ports. Can we just accomplish this by directly committing it to the port allocator ?

Reply:
I see what you mean, no intermediate proposal, just commit directly to the Allocator's `ports` map. I could change it, I guess. Doesn't save a whole lot of time on an already quite expensive operation (Restore of the whole allocator state), though.

Reply:
sure. But low hanging fruit. But I'll let you decide :)